### DISCUSSION TITLE
do_xcom_push doesn't work

### DISCUSSION THREAD

**Discussion by ashwintastic** [apache/airflow#37723] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.5

### What happened?

    cmd2 = 'echo 2'
    dbt_task = BashOperator(task_id='some_task_id',
                            bash_command=cmd2,
                            do_xcom_push=True)

Expectation: 2 should be push to xcom , but nothing is pushed in xcom

### How to reproduce

run a bash command in airflow , with the help of BashOperator with do_xcom_push= True . 

    cmd2 = 'echo 2'
    dbt_task = BashOperator(task_id='some_task_0',
                            bash_command=cmd2,
                            do_xcom_push=True)

### Operating System

windows 11

### Deployment

Docker-Compose

### Anything else?

NA

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by tirkarthi** [comment]
I am unable to reproduce this with below dag which on execution pushes 2 to xcom from some_task_0 task. Can you please add a simple standalone dag without third party dependencies to reproduce this issue? Can you please check logs to see if the bash command was executed to show output in logs?

https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/operators/bash/index.html#airflow.operators.bash.BashOperator

> If BaseOperator.do_xcom_push is True, the last line written to stdout will also be pushed to an XCom when the bash command completes

Sample log : 

```
[2024-02-26, 14:45:32 UTC] {subprocess.py:63} INFO - Tmp dir root location: /tmp
[2024-02-26, 14:45:32 UTC] {subprocess.py:75} INFO - Running command: ['/usr/bin/bash', '-c', 'echo 2']
[2024-02-26, 14:45:32 UTC] {subprocess.py:86} INFO - Output:
[2024-02-26, 14:45:32 UTC] {subprocess.py:93} INFO - 2
[2024-02-26, 14:45:32 UTC] {subprocess.py:97} INFO - Command exited with return code 0
```

```python
from datetime import datetime, timedelta

from airflow import DAG
from airflow.operators.bash import BashOperator

from datetime import timedelta

with DAG(
    dag_id="gh_37709",
    start_date=datetime(2023, 10, 10),
    catchup=False,
    schedule_interval="@daily",
) as dag:

    cmd2 = "echo 2"

    dbt_task = BashOperator(task_id="some_task_0", bash_command=cmd2, do_xcom_push=True)

    dbt_task
```

**Comment by ashwintastic** [comment]
@tirkarthi    Please find my logs
<img width="512" alt="image" src="https://github.com/apache/airflow/assets/11703081/04f7d66c-21cd-4b32-bd5a-8588f7d96bd0">


But XCOM is empty

<img width="580" alt="image" src="https://github.com/apache/airflow/assets/11703081/d277bd82-8c95-4b8c-b846-f3b50790c5e3">

**Comment by potiuk** [comment]
Converting to a disussion. This is more appropriate. @ashwintastic - explaining how exactly you use this BashOperator would be a good idea to get helped and discussion is a better way for that.

**Comment by amoghrajesh** [comment]
@ashwintastic I think you are doing something wrong when it comes to your DAG. I was able to create a DAG against 2.5.0 Airflow and the xcoms seem to work for me. 

Example of the DAG:
```
from datetime import datetime, timedelta
from airflow import DAG
from airflow.operators.bash_operator import BashOperator

default_args = {
    'owner': 'airflow',
    'depends_on_past': False,
    'start_date': datetime(2024, 2, 26),
    'catchup': False,
}

dag = DAG(
    'example_dag_with_bash_operator',
    default_args=default_args,
    description='A simple DAG with BashOperator and XCom push',
    schedule='@once'
)

push_task = BashOperator(
    task_id='push_task',
    bash_command='echo "pushing 2"',
    do_xcom_push=True,
    dag=dag,
)

pull_task = BashOperator(
    task_id='pull_task',
    bash_command='echo "{{ ti.xcom_pull(task_ids=\'push_task\') }}"',
    dag=dag,
)

push_task >> pull_task

```

XCOM pushed:
![image](https://github.com/apache/airflow/assets/35884252/74249500-5646-4d1d-8f3e-3b8e91cb6a9f)


Output of task-2 which is basically pulling xcom:
![image](https://github.com/apache/airflow/assets/35884252/2ddeeb1b-93fb-461a-af81-5de5ea7c18f8)

**Comment by ashwintastic** [comment]
This is my code 

    from airflow.decorators import task
    from airflow.operators.bash import BashOperator

    `@dag(default_args=default_args,
     catchup=False, tags=["abc"],
     schedule_interval=None
     )
    def some_dag():
        databricks_hook = DatabricksHook(databricks_conn_id="some_databricks _id")
        xyz = cu.some_dummy_task()
  





    # cu is some module which has some_dummy_task method

    @task(task_id='some_id', provide_context=True)
    def some_dummy_task(**context):
        ti = context['ti']
       cmd2 = 'echo "pushing 2"'
       dbt_task = BashOperator(task_id='some_dummy_task_id',
                            bash_command=cmd2,
                            do_xcom_push=True)


        dbt_task.execute(context)`

**Reply by potiuk** [reply]
Using operator in decorated task is wrong Where did you have idea to do it that way  ? Was it a Chat-GPT suggested appraoch (because it makes no sense whatsoever) - operator is there to instantiate a task inside a dag, but using operator to run bash task inside a decorated task is well.... pretty nonsense.

**Reply by potiuk** [reply]
Just to give more explanation: If you want to mix classic and task flow operators do it as described here https://airflow.apache.org/docs/apache-airflow/stable/tutorial/taskflow.html#adding-dependencies-between-decorated-and-traditional-tasks 

What you did - you did not create a new task with the BashOperator instantiated and their execute method caled - you just literally called execute method on it and it's not nearly enough to do all the stuff that Airflow Framework does before and after execute.

**Comment by ashwintastic** [comment]
Got it what I was doing wrong. Now Xcom is getting value , but the log is what airflow is printing 

<img width="740" alt="image" src="https://github.com/apache/airflow/assets/11703081/b69b4942-d289-4d7e-b042-b092f72d80d7">


XCOM value

<img width="716" alt="image" src="https://github.com/apache/airflow/assets/11703081/eb48da3a-d37e-4e8d-91e4-e79397fe38a7">

**Comment by potiuk** [comment]
If you have problem with something different. Pleas open a new discussion and ask things there. I have no idea what you are asking about and whether you need any help, but definitely if you do need help explain what you expect to see, how things work and all the circumstances in a *separate* discussion. just throwing a screenshot without explaining what you are asking for does not make it easy for others to help you.

And you should do everything you can to make it easy - because you are asking here for peoples time that they spend in their private time trying to help people like you  who want to use software they got absolutely for free.

So my suggestion is - help others to help you by spending some time on describing your problem. Already several people spend some time at looking at your first problem - and lost more than needed because you have not explained your problem in enough details to help you efficiently.

I suggest you learn from that lesson and be a bit more empathetic towards others and value their time more than yours when all you do is asking for help.

**Comment by ashwintastic** [comment]
@potiuk 

 **Current do_xcom_push behaviour:**:  The last line written to stdout would be pushed to xcom.

 Consider calling a wrapper bash script to execute your bash script and printing the output to stdout. However, the wrapper bash script will also print out the statement that the **execution has been completed**, which would result in the expected output not being pushed to XCOM. According to the screenshot that I have pasted, the highlighted line is printed by the wrapper bash script, which is picked up by the XCOM, but the expectation was **to push ['q', 'h'] to the XCOM**.

**Reply by potiuk** [reply]
@ashwintastic  I am not sure if you read above that you should start a new discssion about it ? And then maybe someone will be able to answer. Tagging me when you ask for help (and continuing unrelated threads) limits your possibilities on getting help
